[MoE] Add deduplicated token dispatch

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #168047
* #167792
* #167676
* #167144
* #166896
* #166899
* #166898
* __->__ #166774
* #163815
* #163814
* #166772
* #165482
* #165101
* #164181
* #164164

Added a `dedup_dispatch` function in `torch.nn.parallell.moe`.

The `dedup_dispatch` performs an inter-node token exchange and top_k index exchange, followed by an intra-node relay to send the token to final destination -- expert -- per the top_k indices.

At the inter-node part, the function performs:
- `make_a2a_exchange_plan`
- `all_to_all_v` of tokens
- `all_to_all_v` of topk indices

At the intra-node part, the function performs:
- a `histc` to convert the received topk indices into token splits
- `make_a2a_2d_exchange_plan`
- get "occurrence number" of each token, an "occurrence number" is the position of a token in an expert chunk from a source rank to a destination rank
- `_all_to_all_v_2d_index_push` of tokens that puts token to the final destination

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

kwen2501 committed:
Update

[ghstack-poisoned]

Skylion007 commented on torch/csrc/distributed/c10d/symm_mem/nvshmem_extension.cu:
```diff
@@ -416,33 +434,32 @@ void _all_to_all_get_inner(
-    args.push_back(b_head_ptr);
   }
 
+  void* args[] = {
```
Nit should be an std::array.

Skylion007 commented on torch/csrc/distributed/c10d/symm_mem/nvshmem_extension.cu:
```diff
@@ -416,33 +434,32 @@ void _all_to_all_get_inner(
   // Optional args for signals
-  void *b_start_ptr, *b_len_ptr, *b_head_ptr;
+  void *b_start_ptr = nullptr, *b_len_ptr = nullptr, *b_head_ptr = nullptr;
   if (has_signal) {
```
Should at least have a proper assert error message check that has_signal implies all the optional args are not null. Ideally a ValueError error message.

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update on "[MoE] Add deduplicated token dispatch"


Added a `dedup_dispatch` function in `torch.nn.parallell.moe`.

The `dedup_dispatch` performs an inter-node token exchange and top_k index exchange, followed by an intra-node relay to send the token to final destination -- expert -- per the top_k indices.

At the inter-node part, the function performs:
- `make_a2a_exchange_plan`
- `all_to_all_v` of tokens
- `all_to_all_v` of topk indices

At the intra-node part, the function performs:
- a `histc` to convert the received topk indices into token splits
- `make_a2a_2d_exchange_plan`
- get "occurrence number" of each token, an "occurrence number" is the position of a token in an expert chunk from a source rank to a destination rank
- `_all_to_all_v_2d_index_push` of tokens that puts token to the final destination

[ghstack-poisoned]
